plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt

```diff
@@ -24,52 +19,45 @@ import org.jetbrains.kotlin.formver.viper.errors.VerificationError
 import org.jetbrains.kotlin.formver.viper.errors.VerifierError
 import org.jetbrains.kotlin.formver.viper.errors.getInfoOrNull
 
-class VerifierErrorInterpreter {
-    private var isOriginalViperErrorReported = false
+object VerifierErrorInterpreter {
 
-    private fun DiagnosticReporter.reportVerificationErrorUserFriendly(
-        source: KtSourceElement?,
-        error: VerificationError,
-        context: CheckerContext,
-    ) {
+    private fun fetchVerificationErrorReporter(error: VerificationError): ErrorReporter<*>? =
         when (val role = error.getInfoOrNull<SourceRole>()) {
-            is SourceRole.ReturnsEffect ->
-                reportOn(source, PluginErrors.UNEXPECTED_RETURNED_VALUE, role.asUserFriendlyMessage, context)
-            is SourceRole.CallsInPlaceEffect ->
-                reportOn(source, PluginErrors.INVALID_INVOCATION_TYPE, role.paramSymbol, role.kind.asUserFriendlyMessage, context)
-            is SourceRole.ParamFunctionLeakageCheck -> with(role) {
-                reportOn(source, PluginErrors.LAMBDA_MAY_LEAK, error.reason.fetchLeakingFunction(), context)
-            }
-            is SourceRole.ConditionalEffect -> {
-                val (returnEffect, condition) = role
-                reportOn(source, PluginErrors.CONDITIONAL_EFFECT_ERROR, returnEffect.describe(), condition.describe(), context)
-            }
-            else -> reportVerificationErrorOriginalViper(source, error, context)
+            is SourceRole.ReturnsEffect -> ReturnsEffectReporter(error, role)
+            is SourceRole.CallsInPlaceEffect -> CallsInPlaceReporter(error, role)
+            is SourceRole.ParamFunctionLeakageCheck -> LeakingLambdaReporter(error, role)
+            is SourceRole.ConditionalEffect -> ConditionalEffectReporter(error, role)
+            else -> null
         }
-    }
 
-    private fun DiagnosticReporter.reportVerificationErrorOriginalViper(
+    private fun DiagnosticReporter.reportOriginalVerificationError(
```

Comment:
Why not make this just another `ErrorReporter`?

Reply:
I thought about this, but then I would kinda be violating the definition of `ErrorReporter` (human-readable) messages:

```kotlin
/**
 * Abstract class defining a [report] method used to warn users with human-readable messages.
 */
abstract class ErrorReporter<S : SourceRole>(
    protected val error: VerificationError,
    protected val role: S,
) 
```

I can implement the default error reporter without problems changing the comment!

Reply:
I don't see any benefit to restricting `ErrorReporter` to only human-readable messages, so yes, I would recommend that.